deepseg.cc

```diff
@@ -464,65 +466,71 @@ int main(int argc, char* argv[]) {
 	cap.set(CV_CAP_PROP_CONVERT_RGB, true);
 
 	calcinfo_t calcinfo = { modelname, threads, width, height, debug };
+	calcinfo.pti = &ti;
 	init_tensorflow(calcinfo);
 
 	// kick off separate grabber thread to keep OpenCV/FFMpeg happy (or it lags badly)
-	pthread_t grabber;
-	cv::Mat buf1;
-	cv::Mat buf2;
-	int64 oldcnt = 0;
-	capinfo_t capinfo = { &cap, &buf1, &buf2, 0, &ti, PTHREAD_MUTEX_INITIALIZER };
-	if (pthread_create(&grabber, NULL, grab_thread, &capinfo)) {
-		perror("creating grabber thread");
-		exit(1);
-	}
+	cv::Mat mask;
+	cv::Mat raw;
 	ti.lastns = timestamp();
 	printf("Startup: %ldns\n", diffnanosecs(ti.lastns,ti.bootns));
 
+	pthread_t ai;
+	if (pthread_create(&ai, NULL, calc_mask_thread, &calcinfo)) {
+		perror("creating calc_mask thread");
+		exit(1);
+	}
+
+
 	bool filterActive = true;
 
 	// mainloop
 	for(bool running = true; running; ) {
-		// wait for next frame
-		while (capinfo.cnt == oldcnt) usleep(10000);
-		oldcnt = capinfo.cnt;
 		int e1 = cv::getTickCount();
-		ti.waitns=timestamp();
 
-		// switch buffer pointers in capture thread
-		pthread_mutex_lock(&capinfo.lock);
+		// grab new frame from cam
+		cap.grab();
+		ti.grabns =  timestamp();
+
+		// copy new frame to buffer
+		cap.retrieve(raw);
+		ti.retrns = timestamp();
+
+		if (raw.rows == 0 || raw.cols == 0) continue; // sanity check
+
+		// copy frame to ai thread local buffer
+		pthread_mutex_lock(&calcinfo.lock_raw);
 		ti.lockns=timestamp();
-		cv::Mat *tmat = capinfo.grab;
-		capinfo.grab = capinfo.raw;
-		capinfo.raw = tmat;
-		pthread_mutex_unlock(&capinfo.lock);
-		// we can now guarantee capinfo.raw will remain unchanged while we process it..
-		calcinfo.raw = *capinfo.raw;
+		calcinfo.raw = raw.clone();
```

Comment:
see above comment regarding double clone, we may want to avoid this one, if there is any material performance improvement.